Create Pronoun model

joshfeli opened the pull request:
This pull request features the `LowercaseCharField` custom field and the `Pronoun` model. The `LowercaseCharField` field inherits from `CharField` but converts the stored string to lowercase upon saving the model instance. The `Pronoun` model is a way to describe single pronouns by defining the word (e.g. `he`) and the pronoun type (e.g. `subj`).

joshfeli committed:
Copied and pasted PronounSeries and Gender classes from GATK into models.py

First step to converting the Gender class into a Django model

joshfeli committed:
Started converting PronounSeries and Gender classes to models

At this point, still figuring out the relationship between the two models, how to implement methods like .lower() in terms of models, and converting more complex Python data types (e.g. sets of PronounSeries instances) to fields

farooqashar committed:
Registering initial models for admin sites

farooqashar committed:
Adding more fields for PronounSeries Model

joshfeli committed:
Refactored model registry in admin.py

Changed variable names to fit conventions of lab

joshfeli committed:
Created custom LowercaseCharField field

Created new field to preserve the case-insensitive properties of the subj and obj attributes of the PronounSeries class (model).  Description for LowercaseCharField is a work in progress (how do we access the max_length?) Also slightly reformatted models_to_register list in admin.py.

joshfeli committed:
Placed custom LowercaseCharField into a newly created fields.py file

Also, slight correction to the attrbitues of the PronounSeries class in models.py.

joshfeli committed:
Created a Pronoun model

The Pronoun model will describe an individual pronoun rather than a set of pronouns. The description will include the type of pronoun (see Pronoun docstring).

joshfeli committed:
Added Pronoun model to registry in admin.py

joshfeli committed:
Isolated changes from create_gender_model branch to Pronoun model

Also left the LowercaseCharField, on which the Pronoun model depends.

joshfeli committed:
Completed a sentence.

joshfeli committed:
Refactored LowercaseCharField to prevent potential aliasing error

joshfeli committed:
Added description to LowercaseCharField

Next steps (probably a new PR): writing tests for Pronoun and LowercaseCharField

joshfeli committed:
Setting up a test suite for Pronoun model; considering a __eq__ method

joshfeli committed:
Setting up a test suite for Pronoun model; considering a __eq__ method

sharabhuiyan commented on backend/app/fields.py:
```diff
@@ -0,0 +1,27 @@
+    """
+
+    # WORK IN PROGRESS!
+    description = gettext_lazy(f"String (up to {super().max_length}) \
```
I think "String (up to length 5) ..." makes more sense.

joshfeli replied:
I've refactored this so that it hopefully reads easier, but why 5 specifically? I decided to use `self.max_length` since the maximum length of the string the `LowercaseCharField` instance represents is determined by the value the user passes in when creating a Model class that uses this field (in my case, the `max_value` is 40, but it may be different if others also decide to use the `LowercaseCharField`).

fyang3 replied:
yes I agree with Joshua! It's more modular in my opinion!

sharabhuiyan commented on backend/app/fields.py:
```diff
@@ -0,0 +1,27 @@
+    description = gettext_lazy(f"String (up to {super().max_length}) \
+            converted to lowercase upon saving the model instance to the database.")
+
+    def pre_save(self, model_instance, add):
```
What is the purpose of the `add` parameter? It is not used in the method.

joshfeli replied:
> What is the purpose of the `add` parameter? It is not used in the method.

The `add` parameter is a Boolean attribute that specifies whether a Model instance should be added to the database or if Django should update a preexisting instance. This overridden `pre_save` method does not use the `add` parameter directly, but the superclass's `pre_save` method includes it and some other subclasses of `CharField` do use the `add` parameter (e.g., `DateField`). I'm looking at [these docs](https://docs.djangoproject.com/en/3.1/ref/models/fields/#django.db.models.Field.pre_save) and [this source code](https://docs.djangoproject.com/en/2.2/_modules/django/db/models/fields/) for reference.

In addition, the `pre_save` method here is one of many methods called in the process of saving a Model instance to the database, and the `add` parameter is passed along. As such, Python raises a `TypeError` when I remove the unused `add` parameter from `pre_save` method.

sharabhuiyan replied:
That sounds good! Thanks for explaining :)

sharabhuiyan commented on backend/app/models.py:
```diff
@@ -1,6 +1,39 @@
+    def __str__(self):
+        pronoun_type_str = ''
+
+        for PRONOUN_TYPE in Pronoun.PRONOUN_TYPES:
```
Is it maybe easier to use a dictionary to represent `PRONOUN_TYPES` ? I feel that it may be easier to find the `pronoun_type_str`.

sharabhuiyan reviewed (changes requested):
Good job! I just commented on a few things that I noticed, please feel free to respond to my comments so that I can approve.

sharabhuiyan commented on backend/app/models.py:
```diff
@@ -1,6 +1,39 @@
+    def __str__(self):
+        pronoun_type_str = ''
+
+        for PRONOUN_TYPE in Pronoun.PRONOUN_TYPES:
```
I'm not sure if this is a better way to do this but you can also do (if it is guaranteed that `self.pronoun_type`  exists in `Pronoun.PRONOUN_TYPES`):
`pronoun_types = dict(Pronoun.PRONOUN_TYPES)`
`pronoun_type_str = pronoun_types[self.pronoun_type]`

MBJean replied:
That's a good suggestion, Shara, though I think Django expects a list of tuples in that `choices` kwarg (per [link](https://docs.djangoproject.com/en/3.2/ref/models/fields/#choices)). Following up on focusing on these lines, though, I wonder if we could use Django's built in `.get_FOO_display()` method here ([docs](https://docs.djangoproject.com/en/3.2/ref/models/instances/#django.db.models.Model.get_FOO_display)).

joshfeli replied:
> That's a good suggestion, Shara, though I think Django expects a list of tuples in that `choices` kwarg (per [link](https://docs.djangoproject.com/en/3.2/ref/models/fields/#choices)). Following up on focusing on these lines, though, I wonder if we could use Django's built in `.get_FOO_display()` method here ([docs](https://docs.djangoproject.com/en/3.2/ref/models/instances/#django.db.models.Model.get_FOO_display)).

Yes, that works well! Thanks for the tip.

MBJean commented on backend/app/models.py:
```diff
@@ -1,6 +1,39 @@
+        ('reflex', 'Reflexive'),
+    ]
+
+    pronoun = LowercaseCharField(max_length=40)
```
This is minor and subjective, but I wonder if this `pronoun` field should be named something other than `pronoun`, since that's also the name of the model. I'm thinking that there might be scenarios that end up looking like this:
```
pronoun = Pronoun(...)
pronoun = pronoun.pronoun
```

Maybe we could call this `identifier` or something?

If we do that, I'd also recommend renaming `pronoun_type` to just `type`. As a matter of personal preference, I like brevity, but YMMV.

joshfeli replied:
Thanks for the suggestion! Would renaming `pronoun_type` to `type` cause aliasing issues with Python's built-in [`type`](https://docs.python.org/3/library/functions.html#type) function?

EDIT: After thinking about it, it probably would not: we're dealing with different scopes here, so `Pronoun.type` should never alias with `builtins.type`.

MBJean commented on backend/app/models.py:
```diff
@@ -1,6 +1,39 @@
+
+        return (
+            f'Pronoun: {self.pronoun}\n\
+            Type: {pronoun_type_str}'
```
I think the multi-line string interpolation here means that the tab characters will also appear. Ex.
```
Pronoun: she
            Type: 
```

MBJean reviewed (changes requested):
Great start, and thanks for the review @sharabhuiyan. Shara and I have both left some comments for consideration, but the changes should all be straightforward to implement. Let me know if you have any questions.

joshfeli committed:
Tweaked overridden description attribute in fields.py

joshfeli committed:
Resolved whitespace issue in models.py

joshfeli committed:
Setting up a test suite for LowercaseCharField

joshfeli committed:
Setting up a test suite for LowercaseCharField

joshfeli committed:
Renamed Pronoun attributes in models.py

joshfeli committed:
Attempting to build some test cases for Pronoun and LowercaseCharField in tests.py

These tests currently don't work (I get this error:  Model class backend.app.models.Pronoun doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.). How do we create good test cases? What should we look to test in models and fields?

joshfeli committed:
Attempting to build some test cases for Pronoun and LowercaseCharField in tests.py

These tests currently don't work (I get this error:  Model class backend.app.models.Pronoun doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.). How do we create good test cases? What should we look to test in models and fields?

joshfeli committed:
Merge branch 'create_pronoun_tests' into create_pronoun_model

joshfeli committed:
Created migrations for changes to models.py

Also deleted the __init__.py file from the backend directory (per Mike's suggestion)

joshfeli committed:
Reformatted 0001_initial.py in migrations module

joshfeli committed:
Applied Pronoun attribute name changes in tests.py

joshfeli committed:
Took advantage of Django's built-in .get_FOO_display() in models.py

joshfeli committed:
Expanded tests for Pronoun model; identified errant results

Also added __eq__ dunder method for Pronoun model (pronouns with the same identifier are classified as equal)

joshfeli committed:
Completed PronounTestCase in tests.py

joshfeli committed:
Added a Pronoun serializer in serializers.py

joshfeli committed:
Removed tests for LowercaseCharField

The LowercaseCharField does not override much from the CharField superclass, and I test the pre_save method implicitly in my tests for the Pronoun class (see the examples with 'his' vs 'HIS' in PronounTestCase in tests.py).

joshfeli commented:
Thanks @MBJean and @sharabhuiyan for your reviews! I went through and implemented all changes, and I also added a `PronounTestCase` in `tests.py` as well as a `PronounSerializer` in `serializers.py`.

The only error that I see that remains is an `ImportError` that raises when using the serializer (it has to do with the relative imports problem that @MBJean and I were discussing earlier), which I have shown in the image that I added to the top-level `docs` directory.

joshfeli committed:
Added image to docs to show serializer bug

fyang3 commented on backend/app/tests.py:
```diff
@@ -1,14 +1,49 @@
+        he = Pronoun.objects.get(identifier='he')
+        self.assertEqual(str(he), 'Pronoun: he\nType: Subject')
+        self.assertEqual(type(he.type), str)
+        he.save()
```
wonder why are you using .save() in here and above - is it possible to omit the method since I don't see a super strong need to touch the database? at least I didn't see he being used multiple times throughout, and perhaps it'll make the test more concise if we don't have the save() but please correct me if you have other ideas in mind!

MBJean replied:
I agree that the `.save()` call here is unnecessary, unless I'm missing something obvious, since we assert that the custom field class is being utilized correctly during `.save()` below with the `his` `Pronoun` instance.

fyang3 commented on backend/app/tests.py:
```diff
@@ -1,14 +1,49 @@
+
+        his = Pronoun.objects.get(identifier='his', type='pos_det')
+        his_caps_until_saving = Pronoun(identifier='HIS', type='pos_pro')
+        self.assertNotEqual(his, his_caps_until_saving)
```
in here, are you trying to test whether the program will save the case insensitive version into your database? Really thoughtful, but I think perhaps what we care more is the result after we saved them since the result is what we will be using, so perhaps the notequal is unnecessary?

MBJean replied:
I like the impulse to shed unnecessary code. In this case, I think I'd argue that we keep the `assertNotEqual` check, as it ensures that the custom field behavior is working as intended.

fyang3 reviewed (commented):
nice work! Only left a couple of comments for Mike and Shara already left many useful notes for you!

MBJean commented on backend/app/tests.py:
```diff
@@ -1,14 +1,49 @@
+        his_caps_until_saving = Pronoun(identifier='HIS', type='pos_pro')
+        self.assertNotEqual(his, his_caps_until_saving)
+
+        his.save()
```
I think that this `.save()` call on `his` is unnecessary, right? The `.create()` call in `setUp` has already performed the `INSERT` statement.

joshfeli replied:
Ah, thank you! I didn't realize the `.create()` call creates and saves an instance all in one--thanks for the heads-up!

MBJean reviewed (approved):
@joshfeli and team. Great work, overall! I've left a few more comments on some code cleanup you can probably undertake, but will leave my approval here. Do take a look at Funing and my's last few comments and let me know if you have any questions. In the meantime, I'll look into the issue with the seralizer.

joshfeli committed:
Removed redundant calls to save() in tests.py

due to a misunderstanding of how the create() method works (it creates and saves a model instance all in one)

MBJean commented:
This LGTM. Thanks for the work! I'll merge it now.
